Implement support for FSAA in combination with post-processing

grorp opened the pull request:
fixes #14285, minetest/irrlicht#286

- Actually it's MSAA I think, or perhaps the terms are equivalent
- I've made it fit into the existing Irrlicht architecture, but that has resulted in code duplication compared to my original "hacky" approach
- OpenGL 3.2+ and OpenGL ES 3.1+ are supported
- EDT_OPENGL3 is not required, EDT_OPENGL works too
- Helpful tutorial: https://learnopengl.com/Advanced-OpenGL/Anti-Aliasing, section "Off-screen MSAA"
- This may be rough around the edges, but in general it works

Screenshot showing it in action:

"master" branch: post-processing, volumetric lighting, FSAA enabled but doesn't work
<img alt="screenshot 1" src="https://github.com/user-attachments/assets/9fbf0910-05fd-4457-a2b2-2ea08f04a862" width="512" />

this PR: post-processing, volumetric lighting, FSAA works
<img alt="screenshot 2" src="https://github.com/user-attachments/assets/d37d8faa-7bb9-4e91-848a-a8d97714f704" width="512" />

P.S. I have used "ChatGPT" as a source of information/answers while working on this

## To do

This PR is a Ready for Review.

- [x] ~~GL error when FSAA is enabled without restart with "opengl" video driver?~~ fixed by 0d3a93d9f251f467845e3440eb45f84845245944 in this PR
- [x] ~~GL warning "framebuffer object is unsupported because the depth/stencil attachment is mismatched" with post-processing with "opengl3" video driver (FSAA state and/or FSAA state on startup may matter)~~ already happens before this PR, "fixed" in PR #15407 (the code triggering the warning was only used due to the workaround, no its no longer used)
- [x] ~~black screen with post-processing with "ogles2" video driver on desktop~~ already happens before this PR, fixed in PR #15396

## How to test

Enable post-processing and FSAA, see that FSAA works.

lhofhansl commented:
oh nice. Going to try today.

lhofhansl commented:
Mostly works, but weird flashes as you move around.

https://github.com/user-attachments/assets/978a8544-5fcc-414c-9fb0-08f9ad6f387e

lhofhansl commented:
Mostly works, but some weird flashes as you move around.

https://github.com/user-attachments/assets/ca0c60c0-5668-48ac-bbbe-a4f4009c416c

lhofhansl commented:
I'll note that now it makes sense to enabled MSAA (for geometry aliasing) *and* FXAA (for texture aliasing)

lhofhansl commented:
The code looks "right". I had started down that rabbit hole a few months ago, and about 70% as far as you got (all the enum/API changes in Irrlicht).

grorp commented:
> Mostly works, but some weird flashes as you move around.
> Happens less the "better" the MSAA. With 16 I do not see any flashes, with 8 I see some, and 4 and 2 I see more. So there's some error that the bloom picks up on.

Can reproduce. With some testing code to compare the bloom texture and the color texture (2868e9264b1d55c145bf538ca0e9981a8847380f):

1st screenshot color texture, 2nd screenshot bloom texture, 3rd screenshot final rendered result
<img alt="color texture" src="https://github.com/user-attachments/assets/013c91ba-147c-4ff1-8075-0e2fa602458d" width="256" /> <img alt="bloom texture" src="https://github.com/user-attachments/assets/7d8ed7b7-36ee-4083-ab1f-68f9441cf5d3" width="256" /> <img alt="final rendered result" src="https://github.com/user-attachments/assets/9cecd4e2-c644-475c-a4ab-acf4b36859c1" width="256" />

If you zoom in, you can see that there are a few #ffffff and #00000 pixels sprinkled across the color texture. The bloom then goes crazy on the #ffffff pixels. I have no idea 1) where the #ffffff pixels are coming from or 2) why the bloom reacts like that

lhofhansl commented:
> If you zoom in, you can see that there are a few #ffffff and #00000 pixels sprinkled across the color texture.

I see nothing wrong with the code, especially the blitting. I tried gl.LINEAR (with only the color texture), but the effect remains the same. It's almost like there's something wrong with the sampling. I wonder if these black and white pixels are there without PP as well, or the way MSAA is enabled with GLX, we just do not notice them.

lhofhansl commented:
Also... When you use SDL (which does not support start-up MSAA in Irrlicht), MSAA now depends on PostProcessing.

lhofhansl commented:
One more. I tried the on-start MSAA with GLX... It produces the same weird pixels.
So I really think these effects were there all along, we just do not notice them without bloom.

grorp commented:
> Also... When you use SDL (which does not support start-up MSAA in Irrlicht), MSAA now depends on PostProcessing.

SDL with normal MSAA works fine for me, also on the "master" branch (obviously only as long as post-processing is disabled, 3d_mode is "none" and undersampling is 0)

Fedora Workstation 40, "NVIDIA GeForce GTX 1080 Ti"

grorp commented on src/client/renderingengine.cpp:
```diff
@@ -179,7 +179,10 @@ RenderingEngine::RenderingEngine(MyEventReceiver *receiver)
+	// Don't enable MSAA in OpenGL context creation if post-processing is enabled,
+	// the post-processing pipeline handles it.
+	bool enable_fsaa = g_settings->get("antialiasing") == "fsaa" &&
+			!g_settings->getBool("enable_post_processing");
```
This has the nice side effect that you can set really high MSAA values without getting blurry fonts or GUIs

lhofhansl replied:
Indeed!

grorp commented:
> Make MSAA part of PP in the settings.

No, MSAA is not inherently a part of the post-processing pipeline nor does it belong into the "Effects" section. It not working without post-processing on your setup doesn't justify moving it into the post-processing section.

> Allow enabled MSAA and FXAA.

Sounds reasonable in general, but not something I want to do in this PR (scope)

grorp commented:
- Fixed the remaining bugs with the legacy "opengl" driver (GL errors; FSAA only working partially if not enabled at startup)
- Fixed the remaining GL errors/warnings: #15396, #15407
- Properly separated FSAA via OpenGL context and via post-processing, only of them is used (this has the nice side effect that you can set really high MSAA values without getting blurry fonts or GUIs when post-processing is enabled)

This PR should now be ready for review.

lhofhansl commented:
The bloom flashing is still there. Should we document that bloom is incompatible with MSAA (in settings other than 16)?

grorp commented:
> The bloom flashing is still there. Should we document that bloom is incompatible with MSAA (in settings other than 16)?

Sure, can't hurt to document it. 4d4e4556e9bd2470510ec16b8403adc355a7561e

lhofhansl reviewed (approved):
Looks good.

As I said I got about 70% there (without ChatGPT :) ), and had made most of the same interface changes.
The main step I was missing how to do the actual resolve myself.

And works fine on my box (with Intel or NVidia drivers)

grorp commented:
> As I said I got about 70% there (without ChatGPT :) ), and had made most of the same interface changes.
> The main step I was missing how to do the actual resolve myself.

The last 10% take 90% of the time, they say :)

> And works fine on my box (with Intel or NVidia drivers)

Nice, thanks for testing.

sfan5 commented:
> If you zoom in, you can see that there are a few #ffffff and #00000 pixels sprinkled across the color texture. The bloom then goes crazy on the #ffffff pixels. I have no idea 1) where the #ffffff pixels are coming from or 2) why the bloom reacts like that

I happened to read https://learnopengl.com/Guest-Articles/2022/Phys.-Based-Bloom yesterday.
You can find a solution in the "Bonus material 0: Graphical debugging" section.

lhofhansl commented:
We have bright and dark pixels. The dark one are just not amplified by bloom.

My vote would be to merge this as.
Then in separate PRs:
- allow MSAA and FXAA enabled. Together they sound be close SSAA but much cheaper
- attempt to fix the bright/dark spots

grorp commented:
> This is ready to merge, right? Just making sure nobody is waiting for anything.

sfan5 self-requested a review earlier, he might still want to review (?)

lhofhansl commented:
btw. if you clamp the color to 0, 1 in the extract_bloom fragment shader, the flashes go away.

Like so:
```
diff --git a/client/shaders/extract_bloom/opengl_fragment.glsl b/client/shaders/extract_bloom/opengl_fragment.glsl
index 36671b06c..1d2563c23 100644
--- a/client/shaders/extract_bloom/opengl_fragment.glsl
+++ b/client/shaders/extract_bloom/opengl_fragment.glsl
@@ -31,5 +31,5 @@ void main(void)
color *= exposure;
#endif

-       gl_FragColor = vec4(color, 1.0); // force full alpha to avoid holes in the image.
+       gl_FragColor = clamp(vec4(color, 1.0), 0.0, 1.0); // force full alpha to avoid holes in the image.
}
```
It's a hack, but gets rid of the flashing. And I did not see any other problems.

@grorp @sfan5

grorp commented on builtin/settingtypes.txt:
```diff
@@ -624,6 +629,10 @@ debanding (Enable Debanding) bool true
 #
+#    Bloom currently causes artifacts if FSAA is enabled and the configured
+#    anti-aliasing scale is low. If you enable bloom and FSAA, make sure to
+#    use a high anti-aliasing scale value.
```
alright, this was suggested by lhofhansl in https://github.com/minetest/minetest/pull/15392#issuecomment-2469103092

grorp commented on irr/src/OpenGL/Driver.cpp:
```diff
@@ -1325,6 +1325,8 @@ void COpenGL3DriverBase::setBasicRenderStates(const SMaterial &material, const S
 
 	// Anti aliasing
+	// Enable MSAA even if it's not enabled in the OpenGL context, we might be
+	// rendering to an FBO with multisampling.
```
it's here to avoid someone putting the code block below behind an `if (AntiAlias >= 2)` or similar in the future

grorp replied:
> also the comment is simply wrong, we are not "enabling msaa" here.

I've already corrected that in [b96319587348f8c1ff5e4573de6ec28abe7d1c5c](https://github.com/minetest/minetest/pull/15392/commits/b96319587348f8c1ff5e4573de6ec28abe7d1c5c)

lhofhansl commented:
@grorp did you see my comment about avoiding the flashing?
It just clamps the light to valid values.
I have tried various settings and have not seen any differences. (Other that the flashing going away).

grorp commented on irr/src/COpenGLCoreRenderTarget.h:
```diff
@@ -172,7 +173,20 @@ class COpenGLCoreRenderTarget : public IRenderTarget
+							textarget = GL_TEXTURE_CUBE_MAP_POSITIVE_X + (int)CubeSurfaces[i];
+							break;
+						default:
+							throw std::logic_error("not reachable");
```
done where it doesn't result in compiler warnings

grorp commented on src/client/render/secondstage.cpp:
```diff
@@ -238,3 +277,14 @@ RenderStep *addPostProcessing(RenderPipeline *pipeline, RenderStep *previousStep
+	GL.BindFramebuffer(GL.DRAW_FRAMEBUFFER, target_fbo_id);
+	v2u32 size = context.target_size;
+	GL.BlitFramebuffer(0, 0, size.X, size.Y, 0, 0, size.X, size.Y,
+			GL.COLOR_BUFFER_BIT | GL.DEPTH_BUFFER_BIT | GL.STENCIL_BUFFER_BIT, GL.NEAREST);
```
Gave it a try. Advantages:

- Somewhat fits into the existing architecture instead of weirdly working around it

Disadvantages:

- It's harder to follow and understand the code because of the interface stuff
- Code duplication between old and new driver
- More code, even without duplication
- When doing this inline, I could be sure that GL version requirements would be met, since I knew the code would only run on GL >= 3.2 (where multisample textures are supported) anyway. Now I can't be sure and have to do extra GL version checks.

sfan5 commented on src/client/render/secondstage.cpp:
```diff
@@ -102,21 +103,52 @@ RenderStep *addPostProcessing(RenderPipeline *pipeline, RenderStep *previousStep
+	// Older versions may support it via the GL_ARB_texture_multisample extension.
+
+	// Note: No support for OpenGL ES implemented. That seems to be another can of worms,
+	// different version requirements, different API functions and possibly some extensions.
```
looks easy actually
https://docs.gl/es3/glTexStorage2DMultisample

grorp replied:
hmm yes. I had looked for gles 2, but with gles 3.1+ it's easy.

enabling it results in a huge performance reduction on my phone though.

lhofhansl commented:
I've noticed those white spots before (without this PR) with flat surfaces, I do not think that's new.

grorp commented on irr/src/COpenGLDriver.cpp:
```diff
@@ -651,6 +651,28 @@ IRenderTarget *COpenGLDriver::addRenderTarget()
+			0, 0, dst->getSize().Width, dst->getSize().Height,
+			GL.COLOR_BUFFER_BIT | GL.DEPTH_BUFFER_BIT | GL.STENCIL_BUFFER_BIT, GL.NEAREST);
+
+	CacheHandler->setFBO(prev_fbo_id, true);
```
OK

this code is duplicated now, yay

sfan5 reviewed (approved):
works

grorp commented:
Oh yay, on my phone it's broken when MSAA is enabled and I remove this code:

https://github.com/minetest/minetest/blob/8d2e7703619205dd76ee9cc12fb0b660868c910b/src/client/render/plain.cpp#L158-L159

with

```
ERROR[Main]: Irrlicht: GL: 8246 824c internalformat 32993 is not a sized format
ERROR[Main]: Irrlicht: GL_INVALID_ENUM COpenGLCoreTexture.h:240
ERROR[Main]: Irrlicht: COpenGLCoreTexture: InternalFormat:0x80e1 PixelFormat:0x80e1
```

It works again if I remove this code too:

https://github.com/minetest/minetest/blob/8d2e7703619205dd76ee9cc12fb0b660868c910b/irr/src/OpenGLES2/Driver.cpp#L89-L92

and results in a significant performance improvement over the current version of this PR. Anyway, not something I want to investigate now.

grorp commented:
> https://registry.khronos.org/OpenGL/extensions/EXT/EXT_texture_format_BGRA8888.txt: `BGRA_EXT` (= the `GL_BGRA` constant we're using) is not sized `BGRA8_EXT` would be

```diff
diff --git a/irr/src/OpenGLES2/Driver.cpp b/irr/src/OpenGLES2/Driver.cpp
index 3c034522c..f33bc61d1 100644
--- a/irr/src/OpenGLES2/Driver.cpp
+++ b/irr/src/OpenGLES2/Driver.cpp
@@ -64,7 +64,7 @@ void COpenGLES2Driver::initFeatures()
                TextureFormats[ECF_D24S8] = {GL_DEPTH24_STENCIL8, GL_DEPTH_STENCIL, GL_UNSIGNED_INT_24_8};
 
                if (FeatureAvailable[IRR_GL_EXT_texture_format_BGRA8888])
-                       TextureFormats[ECF_A8R8G8B8] = {GL_BGRA, GL_BGRA, GL_UNSIGNED_BYTE};
+                       TextureFormats[ECF_A8R8G8B8] = {BGRA8_EXT, GL_BGRA, GL_UNSIGNED_BYTE};
                else if (FeatureAvailable[IRR_GL_APPLE_texture_format_BGRA8888])
                        TextureFormats[ECF_A8R8G8B8] = {BGRA8_EXT, GL_BGRA, GL_UNSIGNED_BYTE};
 
```

or

```diff
diff --git a/irr/src/OpenGLES2/Driver.cpp b/irr/src/OpenGLES2/Driver.cpp
index 3c034522c..9f212b57d 100644
--- a/irr/src/OpenGLES2/Driver.cpp
+++ b/irr/src/OpenGLES2/Driver.cpp
@@ -64,7 +64,7 @@ void COpenGLES2Driver::initFeatures()
                TextureFormats[ECF_D24S8] = {GL_DEPTH24_STENCIL8, GL_DEPTH_STENCIL, GL_UNSIGNED_INT_24_8};
 
                if (FeatureAvailable[IRR_GL_EXT_texture_format_BGRA8888])
-                       TextureFormats[ECF_A8R8G8B8] = {GL_BGRA, GL_BGRA, GL_UNSIGNED_BYTE};
+                       TextureFormats[ECF_A8R8G8B8] = {GL_BGRA8_EXT, GL_BGRA, GL_UNSIGNED_BYTE};
                else if (FeatureAvailable[IRR_GL_APPLE_texture_format_BGRA8888])
                        TextureFormats[ECF_A8R8G8B8] = {BGRA8_EXT, GL_BGRA, GL_UNSIGNED_BYTE};
 
```

make all textures turn black, if that's what you mean

sfan5 commented:
works on my machine™
looking at the revision history this is a recent addition:
>     1.4,  23/06/2024  Erik Faye-Lund: Add GL_BGRA8_EXT for ES 2.0 and later.

how implementations are supposed to deal with this is a mystery to me
